src/server/lib/middleware/login.ts

```diff
@@ -8,40 +8,46 @@ import { trackMetric } from '@/server/lib/trackMetric';
 import { Metrics } from '@/server/models/Metrics';
 import { getConfiguration } from '@/server/lib/getConfiguration';
 import { logger } from '@/server/lib/logger';
+import { addQueryParamsToPath } from '@/server/lib/queryParams';
+import { ResponseWithRequestState } from '@/server/models/Express';
 
 const profileUrl = getProfileUrl();
 
 export const loginMiddleware = async (
   req: Request,
-  res: Response,
+  res: ResponseWithRequestState,
   next: NextFunction,
 ) => {
   const config = getConfiguration();
   const LOGIN_REDIRECT_URL = config.signInPageUrl;
 
-  const FATAL_ERROR_REDIRECT_URL = `${LOGIN_REDIRECT_URL}?error=signin-error-bad-request`;
-
-  const generateRedirectUrl = (url: string): string => {
-    const divider = url.includes('?') ? '&' : '?';
-    return `${url}${divider}returnUrl=${encodeURIComponent(
-      joinUrl(profileUrl, req.path),
-    )}`;
-  };
-
   const redirectAuth = (auth: IDAPIAuthRedirect) => {
     if (auth.redirect) {
-      const redirect = generateRedirectUrl(auth.redirect.url);
+      const redirect = addQueryParamsToPath(auth.redirect.url, {
+        ...res.locals.queryParams,
+        returnUrl: encodeURIComponent(joinUrl(profileUrl, req.path)),
+      });
       return res.redirect(redirect);
     }
 
-    return res.redirect(generateRedirectUrl(LOGIN_REDIRECT_URL));
+    return res.redirect(
+      addQueryParamsToPath(LOGIN_REDIRECT_URL, {
+        ...res.locals.queryParams,
+        returnUrl: encodeURIComponent(joinUrl(profileUrl, req.path)),
```

Comment:
Is it worth allowing the `addQueryParamsToPath` function to deal with the `encodeURIComponent`, This happens by default with [`qs.stringify` function](https://github.com/ljharb/qs#stringifying)

Reply:
ah ha! good spot, the `encodeURIComponent` does not need to happen here!